### DISCUSSION TITLE
Multiple users/seats

### DISCUSSION THREAD

**Discussion by test2a** [Glench/ExtPay#22] [Ideas]
Can multiple uses/users/seats be detected ? I dont want someone to pay once and then reuse credentials in any number of PCs and browsers. 

My extension is in development now and this has me bugged. Is there a way to detect multiple logins to an email and push them to a maybe higher plan say a teams one?

**Comment by Glench** [comment]
Right now users are limited to 10 installs but I was thinking about making that a number you can customize. What would you want to set it to?

I don't have any plans at the moment to have ExtensionPay support a "seat"-based payment system but that may change if more people ask for it. Please comment here or email me if that's something you're needing and explain your use case!

**Reply by amonam** [reply]
Could you kindly consider adding an option in the dashboard that enables us to easily reduce the number of seats from the default 10 to a lower number?

**Comment by burkybang** [comment]
@Glench What do you consider an install? If one user installs it on one computer and it syncs via their Google account to another computer, is that one install or two? Does a user need to log into Extension Pay on both of those computers or does their login sync via their Google account?

**Comment by Glench** [comment]
That's a good question! I haven't tested it but based on my knowledge that all extension data is synced in Chrome's sync system, that would NOT count as another install — only when the user uses the "login" functionality and clicks the magic link in their email.

**Comment by test2a** [comment]
I am not sure about a specific number. How about you allow users to customize this to fit the "single use license" and "multi user license". 

The google cloud sync is there, same as Firefox sync. I dont use either but can we " check" If a user got an extension via sync or login?

**Comment by Glench** [comment]
> The google cloud sync is there, same as Firefox sync. I dont use either but can we " check" If a user got an extension via sync or login?

As far as I know there is no way to detect when a user has "synced" an extension vs installed it normally.

> I am not sure about a specific number. How about you allow users to customize this to fit the "single use license" and "multi user license".

Well this is part of the ambiguity, right? What constitutes a "single use" given users have multiple devices with sync and also different browser profiles? 

I made ExtensionPay use an email with a limited number of logins since that seems like the best experience for users while also not allowing unlimited abuse. And frankly, only one extension user (out of many tens of thousands) I've seen has ever tried to abuse this system which is why I added a 10-install limit.

In general, I would say creators tend to overestimate the amount of abuse their service will receive, especially in the early stages.

@test2a if this concern is really important to you in the near term, then I'd say you should probably build your own service that fits your needs better. Or if that's too much work you might try using ExtensionPay as-is and see if you run into the issue you're imagining and then switch to using your own system — we support exporting all your user data. In the meantime, I'll start thinking about if it makes sense for ExtensionPay to support this case.

**Comment by ianwill93** [comment]
The idea of customizing user seats sounds good for when a product is aimed at businesses. 

How are users currently handling selling in bulk? Does every employee in a company make a separate account?

**Comment by Glench** [comment]
Not many extensions have had organization deals so far. When it has happened the developer just sends me a message with all the emails to add and I add them for them all at once. The employees then use the "log in" functionality to activate the extension.

If you'd like a more automatic or organization-level billing plan, then please explain either here or in an email your use case, how many users you currently have, and how you'd imagine the functionality working.

**Reply by burkybang** [reply]
My need for this has grown every day for the past 5 years. I built a custom solution for this, but it's only halfway there. The lack of this feature is probably my biggest complaint from customers, and I likely lose deals because of it. My customers are a mix of individual users and large corporations.

**Reply by Glench** [reply]
@burkybang Thanks for letting me know! I sent you an email so we can chat about this privately to come up with a solution since not many other people have requested this functionality.

**Comment by surtic86** [comment]
Hi there,

As in a E-Mail from me to you my Users are running in some Issues since they often Switch / Change the Chrome Profiles (delete Them and create New once).

Well you say you can have X active Sessions. Im fine with that and that should be still a thing so i could create a Plan A with 4 seats and a Plan with like 25 seats. 

But what also should be is that we need to Possibilities to Reset that so all Sessions will get Logged out. Or when you Activate a new Session the oldest get Removed.

It can be like it is now where the Users can't Activate on new Systems.

What do you think? Can you make this? 

This thing is kinda urgent since more and more Users are coming to me with this Issue and i expect it will get more over time soon.

**Reply by luizen** [reply]
> Hi there,
> 
> As in a E-Mail from me to you my Users are running in some Issues since they often Switch / Change the Chrome Profiles (delete Them and create New once).
> 
> Well you say you can have X active Sessions. Im fine with that and that should be still a thing so i could create a Plan A with 4 seats and a Plan with like 25 seats.
> 
> But what also should be is that we need to Possibilities to Reset that so all Sessions will get Logged out. Or when you Activate a new Session the oldest get Removed.
> 
> It can be like it is now where the Users can't Activate on new Systems.
> 
> What do you think? Can you make this?
> 
> This thing is kinda urgent since more and more Users are coming to me with this Issue and i expect it will get more over time soon.

Good point. @Glench does the 10 installations/logins limit have a time span after which it is automatically reset or something like that?
Or is it like @surtic86 mentioned, that when the 10th installation/login with the same account happens, the older ones are logged out? I think that approach would be nice :)

Let's say a given user needs to use different computers every day (maybe in a university lab, for instance) and needs to install the extension every time he/she switches computers. What happens when user tries to login the 10th time? Is it blocked?

**Reply by surtic86** [reply]
When he tries to login more then 10 Times he gets this Message: "This email has been used too many times." . I get more and more Customers with that Problem... it's just not good when a Customer can't Login Anymore. This needs to be fixed.

**Comment by burkybang** [comment]
One of my customers is interested in purchasing licenses in bulk and would like the ability to have an admin manage the licenses such as moving licenses between employees.

**Comment by burkybang** [comment]
I have a suggestion for how to handle multiple logins. It comes from how Apple handles iTunes being authorized on multiple devices.

Apple allows for a maximum of 5 simultaneous authorized devices. If a user tries to authorize a 6th device, the user is prevented and provided with a message and an option.

The user can do one of the following:
- From one of the 5 devices, the user can deauthorize the device to free up a slot
- The user can click a button to deauthorize on all 5 devices at once

Apple separates logging in from authorizing. A user can log into iTunes without authorizing the device. This allows for revoking access on all the other devices.

I think a similar method could be achieved with ExtensionPay like so:

After the user clicks the magic link, the user would be logged into ExtensionPay _only_ (not the extension). Then ExtensionPay can do a check to see if the user would exceed the 5 logins limit with the current login.

- If the user _would not_ exceed the 5 logins limit with the current login, then ExtensionPay can send the message to the extension indicating the user is authorized like normal.
- Otherwise, if the user _would_ exceed the 5 logins limit with the current login, then ExtensionPay can ask the user to log out via one of the 5 devices _or_ click a button to revoke all 5 devices.
  - If the user clicks the revoke button, ExtensionPay can invalidate all 5 API keys for each of the 5 logins, create a new API key for the current login, and then send the message to the extension indicating the user is authorized like normal.
  - If the user would like to log out via one of the 5 devices. Then after the user does that, the user can initiate the login process again from the current device to get a new magic link, and the process starts over from the top.

**Comment by JPaulHendricks** [comment]
If I am understanding this all correctly, the default should be 1 IMO. 

It's not clear to me why a subscription, which any user would assume _allows for their single use only,_ would actually allow 10 people (for the same price). 

Why can't it just detect if a second device attempts to authorize (or actively use the premium features) and then disable the earlier auth/usage? The experience of new/second logins forcing a previous login to logout is quite common. 

One subscription=one user. How can that not be the default? But please clarify for me, if I misunderstand. Thanks.

**Reply by Glench** [reply]
This is not how it works currently but we'll change it to be a little more like you described. The problem is that "disable the earlier auth/usage" is actually a terrible user experience. If one user wants to use the extension on multiple chrome profiles, for example, then if they're limited to 1 active use they have to sign in every time they switch chrome profiles. Most users that pay expect to be able to use their extension's paid functionality however they want. 10 installs is an upper limit meant to limit abuse while also minimizing user frustration (which is much more common than abuse, by the way).

**Reply by jellohouse** [reply]
Maybe we could reduce it to like 3-5 installs max. 
I think it's unlikely that someone would install the same extension more than that within their profiles.
At least personally, the extensions i use are mostly different for each profile.
This would prevent more people sharing the same login.

**Reply by Glench** [reply]
Like I said, 10 is an upper limit and is unlikely to be reached. If we limited to 5 installs total across the entire lifetime of your extension then users would hit it constantly.

**Comment by burkybang** [comment]
My need for this has grown every day for the past 5 years.

**Reply by Glench** [reply]
Hi @burkybang, can you be more specific? Are you talking about billing multiple users at a company, or what this thread seems to be about which is managing a single user's experience of logging in to their account multiple times.

**Reply by burkybang** [reply]
@Glench Sorry, I realize that was confusing.

This comment is what I meant to reply to:

https://github.com/Glench/ExtPay/discussions/22#discussioncomment-2017648

**Reply by Glench** [reply]
No worries. I've barely received this request from anyone besides you, so how about you and I chat privately and see if we can come up with something.
